### PR TITLE
Simplify and rewrite dio tests

### LINKED ISSUE #2814: dio deadlock in CI
from https://ci.vmware.run/vmware/vic/6065

```
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420076b00] read from closed multi-reader, returning EOF" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420076b00] read \"\" from 0 readers (err: &errors.errorString{s:\"EOF\"})" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] created multireader" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] read \"verify base multiwriter function\" from 2 readers (err: <nil>)" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removing reader - currently 2 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removed reader - now 1 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] Reader 0 [0xc42007a070]" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removing reader due to EOF" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removing reader - currently 1 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removed reader - now 0 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] All of the readers returned EOF (1)" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] read \"\" from 1 readers (err: <nil>)" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] Going into sleep with 0 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] read \"verify dynamic remove\" from 1 readers (err: <nil>)" 
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [chan receive]:
testing.(*T).Run(0xc42006e0c0, 0x528b52, 0xe, 0x537990, 0xc42003bc01)
    /usr/local/go/src/testing/testing.go:647 +0x316
testing.RunTests.func1(0xc42006e0c0)
    /usr/local/go/src/testing/testing.go:793 +0x6d
testing.tRunner(0xc42006e0c0, 0xc42003bd88)
    /usr/local/go/src/testing/testing.go:610 +0x81
testing.RunTests(0x5379c0, 0x5b9b40, 0x8, 0x8, 0x40edae)
    /usr/local/go/src/testing/testing.go:799 +0x2f5
testing.(*M).Run(0xc42003bee8, 0x52dd80)
    /usr/local/go/src/testing/testing.go:743 +0x85
main.main()
    github.com/vmware/vic/pkg/dio/_test/_testmain.go:118 +0x1bd

goroutine 30 [semacquire]:
sync.runtime_notifyListWait(0xc420077050, 0xc400000000)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420077040)
    /usr/local/go/src/sync/cond.go:57 +0x80
github.com/vmware/vic/pkg/dio.(*multiReader).Read(0xc420077000, 0xc420097400, 0x15, 0x200, 0x0, 0x0, 0x0)
    github.com/vmware/vic/pkg/dio/_test/_obj_test/reader.go:76 +0x279
io.(*LimitedReader).Read(0xc42007cde0, 0xc420097400, 0x200, 0x200, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:436 +0x6c
bytes.(*Buffer).ReadFrom(0xc4200a4230, 0x5a7480, 0xc42007cde0, 0xc420089cb0, 0x1, 0x18)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200a4230, 0x5a7480, 0xc42007cde0, 0x0, 0x0, 0x0, 0x50c040, 0x7fa8f6d74001, 0xc42007cde0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200a4230, 0x5a7480, 0xc42007cde0, 0x507c80, 0x50b200, 0xc420089d00)
    /usr/local/go/src/io/io.go:360 +0x68
io.CopyN(0x5a7340, 0xc4200a4230, 0x7fa8f6d74078, 0xc420077000, 0x15, 0xc420077000, 0xc4200a4230, 0xc420078cc0)
    /usr/local/go/src/io/io.go:336 +0x8c
github.com/vmware/vic/pkg/dio.TestReadRemove(0xc420078cc0)
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:359 +0x8c8
testing.tRunner(0xc420078cc0, 0x537990)
    /usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
    /usr/local/go/src/testing/testing.go:646 +0x2ec

goroutine 17 [semacquire]:
sync.runtime_notifyListWait(0xc420078040, 0x0)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078030)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078000, 0xc420098020, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a000, 0xc420098020, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc4200820e0, 0x5a74c0, 0xc42007a000, 0xc420025ef8, 0xc420078001, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200820e0, 0x5a74c0, 0xc42007a000, 0x0, 0x0, 0x0, 0xc42007a008, 0x5a7500, 0xc42007a018)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200820e0, 0x5a74c0, 0xc42007a000, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestMultiWrite
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:43 +0x322

goroutine 18 [semacquire]:
sync.runtime_notifyListWait(0xc420078100, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc4200780f0)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc4200780c0, 0xc420098620, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a010, 0xc420098620, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082150, 0x5a74c0, 0xc42007a010, 0xc4200206f8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082150, 0x5a74c0, 0xc42007a010, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082150, 0x5a74c0, 0xc42007a010, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestMultiWrite
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:44 +0x36c

goroutine 20 [semacquire]:
sync.runtime_notifyListWait(0xc420078340, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078330)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078300, 0xc420098c32, 0x5ce, 0x5ce, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a020, 0xc420098c32, 0x5ce, 0x5ce, 0x12, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc4200823f0, 0x5a74c0, 0xc42007a020, 0xc420020ef8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200823f0, 0x5a74c0, 0xc42007a020, 0x0, 0x0, 0x0, 0x485f52, 0x5807d6c3, 0xc41ef17ce3)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200823f0, 0x5a74c0, 0xc42007a020, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteAdd
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:76 +0x2fa

goroutine 21 [semacquire]:
sync.runtime_notifyListWait(0xc420078400, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc4200783f0)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc4200783c0, 0xc420099212, 0x5ee, 0x5ee, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a030, 0xc420099212, 0x5ee, 0x5ee, 0x12, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082460, 0x5a74c0, 0xc42007a030, 0xc4200216f8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082460, 0x5a74c0, 0xc42007a030, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082460, 0x5a74c0, 0xc42007a030, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteAdd
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:77 +0x344

goroutine 23 [semacquire]:
sync.runtime_notifyListWait(0xc420078580, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078570)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078540, 0xc420099835, 0x5cb, 0x5cb, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a040, 0xc420099835, 0x5cb, 0x5cb, 0x15, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082a10, 0x5a74c0, 0xc42007a040, 0xc420021ef8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082a10, 0x5a74c0, 0xc42007a040, 0x0, 0x0, 0x0, 0xc42007a048, 0x5a7500, 0xc42007a058)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082a10, 0x5a74c0, 0xc42007a040, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteRemove
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:131 +0x322

goroutine 24 [semacquire]:
sync.runtime_notifyListWait(0xc420078640, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078630)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078600, 0xc4200a0020, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a050, 0xc4200a0020, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082a80, 0x5a74c0, 0xc42007a050, 0xc4200226f8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082a80, 0x5a74c0, 0xc42007a050, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082a80, 0x5a74c0, 0xc42007a050, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteRemove
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:132 +0x36c

goroutine 34 [chan receive]:
github.com/vmware/vic/pkg/dio.TestReadRemove.func4(0xc420080f40, 0xc420090960, 0x5a8500, 0xc420077000, 0xc4200a4230, 0xc420078cc0)
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:349 +0x78
created by github.com/vmware/vic/pkg/dio.TestReadRemove
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:356 +0x852
FAIL    github.com/vmware/vic/pkg/dio   0.009s
make: *** [test-job-github.com/vmware/vic/pkg] Error 1
make: *** Waiting for unfinished jobs....
Makefile:242: recipe for target 'test-job-github.com/vmware/vic/pkg' failed
```

### LINKED ISSUE #3605: Failure in pkg unit test - recipe for target 'test-job-github.com/vmware/vic/pkg' failed
From build: https://ci.vmware.run/vmware/vic/7552

```
time="2017-01-09T21:44:57Z" level=debug msg="[0xc42000b680] Going into sleep with 0 readers" 
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [chan receive]:
testing.(*T).Run(0xc42006e0c0, 0x528b12, 0xe, 0x537a10, 0xc42003bc01)
	/usr/local/go/src/testing/testing.go:647 +0x316
testing.RunTests.func1(0xc42006e0c0)
	/usr/local/go/src/testing/testing.go:793 +0x6d
testing.tRunner(0xc42006e0c0, 0xc42003bd88)
	/usr/local/go/src/testing/testing.go:610 +0x81
testing.RunTests(0x537a40, 0x5b9b20, 0x8, 0x8, 0x40edae)
	/usr/local/go/src/testing/testing.go:799 +0x2f5
testing.(*M).Run(0xc42003bee8, 0x52dd42)
	/usr/local/go/src/testing/testing.go:743 +0x85
main.main()
	github.com/vmware/vic/pkg/dio/_test/_testmain.go:118 +0x1bd

goroutine 19 [semacquire]:
sync.runtime_notifyListWait(0xc42000b6d0, 0xc400000000)
	/usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc42000b6c0)
	/usr/local/go/src/sync/cond.go:57 +0x80
github.com/vmware/vic/pkg/dio.(*multiReader).Read(0xc42000b680, 0xc4200aa000, 0x15, 0x200, 0x0, 0x0, 0x0)
	github.com/vmware/vic/pkg/dio/_test/_obj_test/reader.go:76 +0x279
io.(*LimitedReader).Read(0xc42009a0c0, 0xc4200aa000, 0x200, 0x200, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:436 +0x6c
bytes.(*Buffer).ReadFrom(0xc42008c310, 0x5a7480, 0xc42009a0c0, 0xc42007bcb0, 0x1, 0x18)
	/usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc42008c310, 0x5a7480, 0xc42009a0c0, 0x0, 0x0, 0x0, 0x50bf40, 0x7fdffd9d2201, 0xc42009a0c0)
	/usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc42008c310, 0x5a7480, 0xc42009a0c0, 0x507c20, 0x50b1a0, 0xc42007bd00)
	/usr/local/go/src/io/io.go:360 +0x68
io.CopyN(0x5a7340, 0xc42008c310, 0x7fdffd9d2290, 0xc42000b680, 0x15, 0xc42000b680, 0xc42008c310, 0xc42006ed80)
	/usr/local/go/src/io/io.go:336 +0x8c
github.com/vmware/vic/pkg/dio.TestReadRemove(0xc42006ed80)
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:359 +0x8c8
testing.tRunner(0xc42006ed80, 0x537a10)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec

goroutine 6 [semacquire]:
sync.runtime_notifyListWait(0xc42006e280, 0x0)
	/usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc42006e270)
	/usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc42006e240, 0xc420082020, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc420028028, 0xc420082020, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc4200161c0, 0x5a74c0, 0xc420028028, 0xc420025ef8, 0xc42006e201, 0x0)
	/usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200161c0, 0x5a74c0, 0xc420028028, 0x0, 0x0, 0x0, 0xc420028030, 0x5a7500, 0xc420028040)
	/usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200161c0, 0x5a74c0, 0xc420028028, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestMultiWrite
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:43 +0x322

goroutine 7 [semacquire]:
sync.runtime_notifyListWait(0xc42006e340, 0x1)
	/usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc42006e330)
	/usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc42006e300, 0xc420082620, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc420028038, 0xc420082620, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420016230, 0x5a74c0, 0xc420028038, 0xc4200266f8, 0x1, 0x0)
	/usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420016230, 0x5a74c0, 0xc420028038, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420016230, 0x5a74c0, 0xc420028038, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestMultiWrite
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:44 +0x36c

goroutine 9 [semacquire]:
sync.runtime_notifyListWait(0xc42006e4c0, 0x1)
	/usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc42006e4b0)
	/usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc42006e480, 0xc420082c32, 0x5ce, 0x5ce, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc420028048, 0xc420082c32, 0x5ce, 0x5ce, 0x12, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc4200164d0, 0x5a74c0, 0xc420028048, 0xc420026ef8, 0x1, 0x0)
	/usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200164d0, 0x5a74c0, 0xc420028048, 0x0, 0x0, 0x0, 0x4860d2, 0x58740459, 0xc421b94b2e)
	/usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200164d0, 0x5a74c0, 0xc420028048, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteAdd
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:76 +0x2fa

goroutine 10 [semacquire]:
sync.runtime_notifyListWait(0xc42006e580, 0x1)
	/usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc42006e570)
	/usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc42006e540, 0xc420083212, 0x5ee, 0x5ee, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc420028058, 0xc420083212, 0x5ee, 0x5ee, 0x12, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420016540, 0x5a74c0, 0xc420028058, 0xc4200276f8, 0x1, 0x0)
	/usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420016540, 0x5a74c0, 0xc420028058, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420016540, 0x5a74c0, 0xc420028058, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteAdd
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:77 +0x344

goroutine 12 [semacquire]:
sync.runtime_notifyListWait(0xc42006e700, 0x1)
	/usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc42006e6f0)
	/usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc42006e6c0, 0xc420083835, 0x5cb, 0x5cb, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc420028068, 0xc420083835, 0x5cb, 0x5cb, 0x15, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420016af0, 0x5a74c0, 0xc420028068, 0xc420027ef8, 0x1, 0x0)
	/usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420016af0, 0x5a74c0, 0xc420028068, 0x0, 0x0, 0x0, 0xc420028070, 0x5a7500, 0xc420028080)
	/usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420016af0, 0x5a74c0, 0xc420028068, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteRemove
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:131 +0x322

goroutine 13 [semacquire]:
sync.runtime_notifyListWait(0xc42006e7c0, 0x1)
	/usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc42006e7b0)
	/usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc42006e780, 0xc420088020, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc420028078, 0xc420088020, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
	/usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420016b60, 0x5a74c0, 0xc420028078, 0xc4200206f8, 0x1, 0x0)
	/usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420016b60, 0x5a74c0, 0xc420028078, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420016b60, 0x5a74c0, 0xc420028078, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteRemove
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:132 +0x36c

goroutine 34 [chan receive]:
github.com/vmware/vic/pkg/dio.TestReadRemove.func4(0xc420013530, 0xc420018cc0, 0x5a8500, 0xc42000b680, 0xc42008c310, 0xc42006ed80)
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:349 +0x78
created by github.com/vmware/vic/pkg/dio.TestReadRemove
	/drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:356 +0x852
FAIL	github.com/vmware/vic/pkg/dio	0.024s
Makefile:256: recipe for target 'test-job-github.com/vmware/vic/pkg' failed
make: *** [test-job-github.com/vmware/vic/pkg] Error 1
make: *** Waiting for unfinished jobs....
```

### PR DESCRIPTION
**Pull request by caglar10ur**
to solve races detected by race detector and CI.

Fixes #3605
Fixes #2814

### CODE REVIEW

**Comment by fdawg4l** [comment]
LGTM but I don't really know this code at all.

File: pkg/dio/multi_test.go [hunk 1]
```diff
@@ -12,365 +12,278 @@
...
-func TestMultiWrite(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+func read(t *testing.T, mreader DynamicMultiReader, limit int) []byte {
+	total := 0
 
-	mwriter := MultiWriter(pipeAW, pipeBW)
+	var buf = make([]byte, 32*1024)
+	for {
+		n, err := mreader.Read(buf)
```
Comment (hickeng) [inline pkg/dio/multi_test.go]: think this is supposed to be a slice:
`buf[total:limit]`

if you don't have the first you're overwriting data, if you don't have the latter you may read past limit in the reader and lose data against a subsequent call.

File: pkg/dio/multi_test.go [hunk 2]
```diff
@@ -12,365 +12,278 @@
...
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+func control(t *testing.T, buffers []*bytes.Buffer, p []byte) {
+	for i := range buffers {
+		buffer := buffers[i]
+		if buffer.String() != string(p) {
```
Comment (hickeng) [inline pkg/dio/multi_test.go]: `bytes.Equal`?

File: pkg/dio/multi_test.go [hunk 3]
```diff
@@ -12,365 +12,278 @@
...
+			t.Errorf("Expected: %q, actual: %q", string(p), buffer.String())
+		}
 	}
+}
 
+func equal(t *testing.T, mreader DynamicMultiReader, str string, count int) {
+	expected := strings.Repeat(str, count)
+	// read from multi reader
+	buffer := string(read(t, mreader, len(expected)))
```
Comment (hickeng) [inline pkg/dio/multi_test.go]: why do we have this and control? should one be calling the other or is this just a string typed variant of control - in which case more consistent function naming please.

File: pkg/dio/multi_test.go [hunk 4]
```diff
@@ -12,365 +12,278 @@
...
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != "" {
-		t.Errorf("B: expected: %s, actual: %s", "", bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
```
Comment (hickeng) [inline pkg/dio/multi_test.go]: how do we ensure that all of the io.Copy's have finished by the time we compare buffers?

File: pkg/dio/multi_test.go [hunk 5]
```diff
@@ -12,365 +12,278 @@
...
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != data2 {
-		t.Errorf("B: expected: %s, actual: %s", data2, bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
```
Comment (hickeng) [inline pkg/dio/multi_test.go]: same comment as above about ensuring completion, and for all subsequent control calls.

File: pkg/dio/multi_test.go [hunk 6]
```diff
@@ -12,365 +12,278 @@
...
 
-	data2 := "verify dynamic remove"
-	_, err = mwriter.Write([]byte(data2))
-	if err != nil {
-		t.Error(err)
-		return
-	}
+	// add the writer & buffer to skip list if i divisible by three
+	for i := 0; i < count; i++ {
```
Comment (hickeng) [inline pkg/dio/multi_test.go]: combine this with the preceding loop so the structure is the same as the Add test.

File: pkg/dio/multi_test.go (hunk 1 above)
Comment (caglar10ur) [inline pkg/dio/multi_test.go]: Done

File: pkg/dio/multi_test.go (hunk 2 above)
Comment (caglar10ur) [inline pkg/dio/multi_test.go]: Started to use assert library

File: pkg/dio/multi_test.go (hunk 3 above)
Comment (caglar10ur) [inline pkg/dio/multi_test.go]: Done (removed)

File: pkg/dio/multi_test.go (hunk 4 above)
Comment (caglar10ur) [inline pkg/dio/multi_test.go]: Done

File: pkg/dio/multi_test.go (hunk 5 above)
Comment (caglar10ur) [inline pkg/dio/multi_test.go]: Done

File: pkg/dio/multi_test.go (hunk 6 above)
Comment (caglar10ur) [inline pkg/dio/multi_test.go]: Done

**Comment by caglar10ur** [comment]
@hickeng diff is crowded - suggest looking at the file directly (https://github.com/caglar10ur/vic/blob/0222b4d4704d4be7c68dabb05e3b5ee211818f5a/pkg/dio/multi_test.go)

**Comment by hickeng** [comment]
@caglar10ur didn't do a detailed review - checked those aspects that were issues before.

**Pull request merged**